infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentDao.kt

```diff
@@ -228,99 +244,121 @@ class LayoutAlignmentDao(jdbcTemplateParam: NamedParameterJdbcTemplate?) : DaoBa
-                    "switch_end_joint_number" to s.endJointNumber?.intValue,
-                    "start" to s.start,
-                    "length" to s.length,
-                    "source_start" to s.sourceStart,
-                    "source" to s.source.name,
+            val newGeometryIds = insertSegmentGeometries(segments.mapNotNull { s ->
+                if (s.geometry.id is StringId) s.geometry else null
+            })
+            //language=SQL
+            val sqlIndexed = """
+                insert into layout.segment_version(
+                  alignment_id,
+                  alignment_version,
+                  segment_index,
+                  geometry_alignment_id,
+                  geometry_element_index,
+                  switch_id,
+                  switch_start_joint_number,
+                  switch_end_joint_number,
+                  source_start,
+                  source,
+                  geometry_id
                 )
-            }.toTypedArray()
-            jdbcTemplate.batchUpdate(sql, params)
+                values(?, ?, ?, ?, ?, ?, ?, ?, ?, ?::layout.geometry_source, ?) 
+              """.trimIndent()
+            jdbcTemplate.batchUpdateIndexed(sqlIndexed, segments) { ps, (index, s) ->
+                ps.setInt(1, alignmentId.id.intValue)
+                ps.setInt(2, alignmentId.version)
+                ps.setInt(3, index)
+                ps.setNullableInt(4) { if (s.sourceId is IndexedId) s.sourceId.parentId else null }
+                ps.setNullableInt(5) { if (s.sourceId is IndexedId) s.sourceId.index else null }
+                ps.setNullableInt(6) { if (s.switchId is IntId) s.switchId.intValue else null }
+                ps.setNullableInt(7, s.startJointNumber?.intValue)
+                ps.setNullableInt(8, s.endJointNumber?.intValue)
+                ps.setNullableDouble(9, s.sourceStart)
+                ps.setString(10, s.source.name)
+                val geometryId =
+                    if (s.geometry.id is IntId) s.geometry.id
+                    else requireNotNull(newGeometryIds[s.geometry.id]) { "SegmentGeometry not stored: id=${s.id}" }
+                ps.setInt(11, geometryId.intValue)
+            }
         }
+    }
 
-        if (alignmentId.version > 1) {
-            val sqlDelete = """ 
-              delete from layout.segment 
-              where alignment_id = :alignment_id 
-                and segment.alignment_version < :alignment_version  
-            """.trimIndent()
-
-            val paramsDelete = mapOf(
-                "alignment_id" to alignmentId.id.intValue,
-                "alignment_version" to alignmentId.version,
+    // TODO: GVT-1691 batching this is a little tricky due to difficulty in mapping generated ids:
+    //  There is no guarantee of result set order (though it's usually the insert order)
+    //  If we could calculate the hash prior to saving we could use that to identify the mapping
+    private fun insertSegmentGeometries(
```

Comment:
Kuten todo sanoo, uudet geometriat insertoidaan nyt tyhmästi yksi kerrallaan, koska en tiennyt varmaa tapaa saada batch insertistä generoituja id:tä ulos niin että voisin antaa ne insert vastaukseen (segmentin foreign key:tä varten). Periaatteessa batch insert kyllä antaa paluuarvot samassa järjestyksessä kun insert tehdään, mutta se ei ole dokumentoitu ominaisuus ja pienellä tonkimisella törmäsin kehittäjien kommentteihin että jatkossa näin ei ehkä enää olisi. Jos järjestykseen ei voi luottaa niin ulos tulevat id:t on hankala mapata olioihin jotka oltiin insertoimassa.

Toisaalta, tämä ei ehkä haittaa niinkään paljoa, sillä yleensä vain melko rajallinen joukko geometrioita oikeasti muuttuu kerrallaan ja muuttumattomien segmenttien kohdalla ei tallenneta mitään. Tästä voi kuitenkin tulla takkuilua split-featuren kanssa, joten jos on hyviä ajatuksia niin otan mielelläni vastaan.